### PR TITLE
Documenting saga diagnostics

### PR DESCRIPTION
**Pull request by DavidBoike**

### CODE REVIEW

File: nservicebus/sagas/analyzers.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+## Message types should not be used as saga data properties
+
+* **Rule ID**: NSB0010
+* **Severity**: Warning
+* **Example message**: Using the message type `MyMessage` to store message contents in saga data is not recommended, as it creates unnecessary coupling between the structure of the message and the stored saga data, making both more difficult to evolve.
+
+When a saga receives a message, it can be convenient and even tempting to insert the whole thing into the saga data.
+
+However, this creates unintended couling between the saga data and the message contract.
```
Comment (cquirosj) [inline nservicebus/sagas/analyzers.md:99]:
```suggestion
However, this creates unintended coupling between the saga data and the message contract.
```

File: nservicebus/sagas/analyzers.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
+---
+title: Roslyn analyzers for sagas
+summary: Details of the Roslyn analyzers that promote code quality in sagas.
+component: Core
+versions: '[7.7,)'
+reviewed: 2022-03-04
+---
+
+Starting in NServiceBus version 7.7, [Roslyn analyzers](https://docs.microsoft.com/en-us/visualstudio/code-quality/roslyn-analyzers-overview) are packaged with the NServiceBus package that analyze the code in sagas and make suggestions for improvements, right in the editor.
```
Comment (cquirosj) [inline nservicebus/sagas/analyzers.md:9]:
```suggestion
Starting in NServiceBus version 7.7, [Roslyn analyzers](https://docs.microsoft.com/en-us/visualstudio/code-quality/roslyn-analyzers-overview) are packaged with the NServiceBus package that analyzes the code in sagas and makes suggestions for improvements, right in the editor.
```

File: nservicebus/sagas/analyzers.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+The generic class constraints on `Saga<TSagaData>` require the saga data class to implement the `IContainSagaData` interface, which specifies properties required by the saga infrastructure. However, it is much easier to directly inherit `ContainSagaData`, which already specifies these properties.
+
+A benefit to inheriting the `ContainSagaData` class is that in NServiceBus version 7 and above the implemented properties are decorated with `[EditorBrowsable(EditorBrowsableState.Never)]`, which means that those properties that are _only_ needed by the saga infrastructure will not appear in IntelliSense. This makes it less likely that one of these reserved properties will be used accidentally.
+
+One exception comes when [using NHibernate's `[RowVersion]` attribute to control optimistic concurrency ](/persistence/nhibernate/saga-concurrency.md#custom-behavior-explicit-version). This attribute is not compatible with derived classes, so in this case the saga data class must implement `IContainSagaData` directly. In this case the `NSB0012` diagnostic can be suppressed to remove the warning.
```
Comment (cquirosj) [inline nservicebus/sagas/analyzers.md:127]:
```suggestion
One exception comes when [using NHibernate's `[RowVersion]` attribute to control optimistic concurrency](/persistence/nhibernate/saga-concurrency.md#custom-behavior-explicit-version). This attribute is not compatible with derived classes, so in this case the saga data class must implement `IContainSagaData` directly. In this case the `NSB0012` diagnostic can be suppressed to remove the warning.
```

**Review by cquirosj — APPROVED** [review]

File: nservicebus/sagas/analyzers.md (hunk 2 above)
Comment (DavidBoike) [reply nservicebus/sagas/analyzers.md:9]: Nope, Roslyn analyzers here is plural, so noun-verb agreement means "analyze" and "make".

**Review by TravisNickels — APPROVED** [review]

**Pull request merged**